Added Opensea's Operator Filter Registry

datboi-1337 opened the pull request:
This makes the necessary changes to the contract to enable Opensea's new creator fee filters. 

This does, of course, restrict the collections using it to only use marketplaces approved by Opensea, which I personally don't love. If we want to be able to set and collect creator fees on Opensea, however, it looks like this is the route we have to take for now. 

Because the contracts provided by Opensea require 0.8.13, the compiler version has been changed here as well, and with that, comes specifying overridden contracts. Apart from that, the filter registry stuff has been imported, inherited and the transfer function overrides have been added.

datboi-1337 committed:
opensea

datboi-1337 committed:
Added Opensea creator fee compliance

Added operator-filter-registry from ProjectOpensea and made necessary changes to facilitate that change. Hardhat compiler version updated to 0.8.13.

AnaTekPrimal commented:
This is gold, thank you.

datboi-1337 committed:
Merge pull request #1 from hashlips-lab/main

Updating dependencies to support NodeJS 18 out-of-the-box (#428)

JohnCarp commented:
Hey @liarco I hope all is well!

Is there any chance you can take a look at this PR and make some suggestions? Opensea added more functions to this registry and when you try to implement it in this project there are a lot of compile errors.

I see there was a lot of discussion in the Hashlips Discord and the community really can use your help!

https://github.com/ProjectOpenSea/operator-filter-registry

```
    function setApprovalForAll(address operator, bool approved) public override onlyAllowedOperatorApproval(operator) {
        super.setApprovalForAll(operator, approved);
    }

    function approve(address operator, uint256 tokenId) public override onlyAllowedOperatorApproval(operator) {
        super.approve(operator, tokenId);
    }

    function transferFrom(address from, address to, uint256 tokenId) public override onlyAllowedOperator(from) {
        super.transferFrom(from, to, tokenId);
    }

    function safeTransferFrom(address from, address to, uint256 tokenId) public override onlyAllowedOperator(from) {
        super.safeTransferFrom(from, to, tokenId);
    }

    function safeTransferFrom(address from, address to, uint256 tokenId, bytes memory data)
        public
        override
        onlyAllowedOperator(from)
    {
        super.safeTransferFrom(from, to, tokenId, data);
    }
```

Thank you

datboi-1337 committed:
Updated for approval functions

Updated contract be fully compliant with Opensea guidelines for the Operator Filter Registry.

Edited mintForAddress to allow owner to mint more than maxPerTx, but will still restrict minting more than maxSupply.

Added separators and reorganized contract a bit to make it a little easier to navigate.

datboi-1337 commented:
I was able to get in here and update everything to be fully in compliance with Opensea's Operator Filter Registry, including the approval functions.

JohnCarp commented on smart-contract/contracts/YourNftToken.sol:
```diff
@@ -68,15 +73,21 @@ contract YourNftToken is ERC721AQueryable, Ownable, ReentrancyGuard {
   }
   
-  function mintForAddress(uint256 _mintAmount, address _receiver) public mintCompliance(_mintAmount) onlyOwner {
+  function mintForAddress(uint256 _mintAmount, address _receiver) public onlyOwner {
```
I'm curious why you removed `mintCompliance` modifier from this function?

Also, if this is ok shouldn't you have the other require there? i.e:

```
function mintForAddress(uint256 _mintAmount, address _receiver) public onlyOwner { 
    require(_mintAmount > 0 && _mintAmount <= maxMintAmountPerTx, 'Invalid mint amount!');
    require(totalSupply() + _mintAmount <= maxSupply, 'Max supply exceeded!');

    _safeMint(_receiver, _mintAmount);
}
```

datboi-1337 replied:
This is unrelated to the Opernsea filter, just a general tweak that I have found people have issues with sometimes. I did briefly mention it in the push, but this change simply removes the restriction from the owner to mint more than the current maxMintAmountPerTx. It still ensures that maxSupply will not be exceeded, but allows the contract owner to mint multiple tokens, even if maxMintAmountPerTx is set to 1, for example.

JohnCarp commented:
That's pretty good! Left a comment @datboi-1337
